### PR TITLE
Add customizations for apps.krusie.io/v1alpha1/CloneSet

### LINKED ISSUE #3331: Add third-party resources into the Resource Interpreter framework
**What would you like to be added**:
Karmada's Resource Interpreter Framework is designed for interpreting resource structure. It consists of built-in and customized interpreters. Karmada could bundle some popular and open-sourced resources so that users can save the effort to customize them.

**Tasks**:
1. Argo Workflow  (deadline: April 23, 2023)
    group: argoproj.io
    version: v1alpha1
    - [x] Implement resoucecustomization for  Workflow, including `InterpretReplica`, `ReviseReplica`, `Retain`,  `InterpretDependency`,`InterpretHealthy`. #3438 

2. Flux CD (deadline: April 30, 2023)
    group: source.toolkit.fluxcd.io
    version: v1
    - [x] implement resoucecustomization for  GitRepository, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3481 

    group: source.toolkit.fluxcd.io
    version: v1beta2
    - [x] implement resoucecustomization for OCIRepository, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3482
    - [x] implement resourcecustomization for HelmRepository, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3484
    - [x] implement resoucecustomization for Bucket, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3485 
    - [x] implement resourcecustomization for HelmChart, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3486 

    group: kustomize.toolkit.fluxcd.io
    version: v1
    - [x] implement resoucecustomization for Kustomization, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3487 

    group: helm.toolkit.fluxcd.io
    version: v2beta1
    - [x] implement resoucecustomization for HelmRelease, including `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`. #3488
    
3. Kyverno (deadline: May 10, 2023)
    group: kyverno.io    
    version: v1
    - [x] Implement resoucecustomization for Policy, including `AggregateStatus`, `InterpretStatus`, `InterpretHealth`. #3498 
    - [x] Implement resoucecustomization for ClusterPolicy, including `AggregateStatus`, `InterpretStatus`, `InterpretHealth`. #3499 
    
4. OpenKruise (deadline: ~April 1, 2023~ April 4, 2023)
    group: apps.kruise.io    
    version: v1alpha1
    - [x] Implement resoucecustomization for CloneSet, including `InterpretReplica`, `ReviseReplica`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`, `InterpretHealthy`.#3337
    - [x] Implement resourcecustomization for AdvancedDaemonSet, including `AggregateStatus`, `InterpretStatus`, `InterpretDependency`, `InterpretHealthy`.#3358
    - [x] Implement resourcecustomization for BroadcastJob, including `InterpretReplica`, `ReviseReplica`, `Retain`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`,`InterpretHealthy`.#3366
    - [x] Implement resourcecustomization for AdvancedCronJob, including `AggregateStatus`, `InterpretDependency`, `InterpretHealthy`.#3368 

    group: apps.kruise.io    
    version: v1beta1
    - [x] Implement resourcecustomization for AdvancedStatefulSet, including `InterpretReplica`, `ReviseReplica`, `AggregateStatus`, `InterpretStatus`, `InterpretDependency`, `InterpretHealthy`.#3352

  Others those should be implemented.
  - [x] #3339
  - [x] update the document. https://github.com/karmada-io/website/pull/360

**Why is this needed**:

### PR DESCRIPTION
**Pull request by yike21**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Add third-party resources `apps.krusie.io/v1alpha1/CloneSet` into Resource Interpreter framework.

**Which issue(s) this PR fixes**:
Part of #3331

**Special notes for your reviewer**:
@Poor12 

**Does this PR introduce a user-facing change?**:
 NONE

```release-note
This is the first resourceinterpretercustomization for krusie.
Maybe it's useful to add a function `getDependenciesFromPodTemplate` in `pkg\resourceinterpreter\configurableinterpreter\luavm\kube.go`.
The directory for those customizatons is pending.
```

### CODE REVIEW

**Comment by XiShanYongYe-Chang** [comment]
> /assign @lonelyCZ

Sorry, I make a mistake.

/unassign @lonelyCZ 
/assign @Poor12

**Comment by Poor12** [comment]
Please use `Part of #3331` instead of `Fixes #3331` because the latter may close the issue.

File: resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+          status.expectedUpdatedReplicas = observedObj.status.expectedUpdatedReplicas
+          status.updateRevision = observedObj.status.updateRevision
+          status.currentRevision = observedObj.status.currentRevision
+          return status
+        end
+    healthInterpretation:
+      luaScript: >
+        function InterpretHealth(observedObj)
+          return observedObj.status.readyReplicas == observedObj.spec.replicas
+          --[[
```
Comment (chaunceyjiang) [inline resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:103]: Can this line be deleted?

Comment (yike21) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:103]: Yes, I will delete it soon.

**Comment by chaunceyjiang** [comment]
> Maybe it's useful to add a function `getDependenciesFromPodTemplate` in `pkg\resourceinterpreter\configurableinterpreter\luavm\kube.go`.


Good idea!

File: resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+    dependencyInterpretation:
+      luaScript: >
+        function GetDependencies(desiredObj)
+          dependentConfigMaps = {}
+          dependentSecrets = {}
+          dependentSas = {}
+          dependentPVCs = {}
+          refs = {}
+          local idx = 1
+          if desiredObj.spec.template.spec.initContainers ~= nil then
```
Comment (chaunceyjiang) [inline resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:127]: As you said, maybe we can introduce a new helper function `getDependencies From PodTemplate` to reduce duplicate code.

Comment (yike21) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:127]: Yes! If we add this function, It's also easy to complete `InterpretDependency` operation using lua, like `InterpretReplica` with `kube.accuratePodRequirements`.

Comment (Poor12) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:127]: I agree with it. Would you like to send a separate PR to improve this? @yike21

Comment (yike21) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:127]: Yes, I'm willing to do it!

Comment (Poor12) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:113]: As #3339 is merged,  we can start to refactor the codes about `interpretDependency`.

Comment (yike21) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:113]: OK, I will refactor the codes soon!

**Comment by Poor12** [comment]
I guess there is something wrong about `interpretDependency`. I use the `customizations_propagationpolicy_test.yaml` but it seems getDependency failed. From the log:
```
  Warning  GetDependenciesFailed   51s (x16 over 3m34s)   dependencies-distributor            <string>:2: attempt to call a non-function object
stack traceback:
  <string>:2: in main chunk
  [G]: ?

```

**Comment by yike21** [comment]
> I guess there is something wrong about `interpretDependency`. I use the `customizations_propagationpolicy_test.yaml` but it seems getDependency failed. From the log:
> 
> ```
>   Warning  GetDependenciesFailed   51s (x16 over 3m34s)   dependencies-distributor            <string>:2: attempt to call a non-function object
> stack traceback:
>   <string>:2: in main chunk
>   [G]: ?
> ```
> 
> Please help check it.
I test this and find the problem again. In the past, I test `interpretDependency` using lua and not kube function, because my local environment is latest karmada-controller image released on Feb 28. It might be because the function `kube.getPodDependencies` is not yet available in the latest image version.

**Comment by chaunceyjiang** [comment]
On my environment it works fine. @Poor12 


image

**Comment by Poor12** [comment]
> On my environment it works fine. @Poor12
> 
> image

Thanks, I have solved this issue. BTW, I think we should provide a document for users about how to contribute third-party well-known CRDs. WDYT @chaunceyjiang

**Comment by chaunceyjiang** [comment]
>  I think we should provide a document for users about how to contribute third-party well-known CRDs. 

+1

File: resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations_propagationpolicy_test.yaml
Comment (chaunceyjiang) [inline resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations_propagationpolicy_test.yaml:1]: This file is just for testing,  so , can we put this file in the `testdata` dir ?

Comment (yike21) [reply resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations_propagationpolicy_test.yaml:1]: It sounds great! These test files are just for making testing and verification more convenient.

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/resourceinterpretercustomization.yaml
Comment (chaunceyjiang) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/resourceinterpretercustomization.yaml:1]: https://github.com/karmada-io/karmada/blob/538d8f10e091bf520581377d3598820fe06b327f/pkg/resourceinterpreter/default/thirdparty/manager.go#L21


oh, now there is a fixed name for `customizations.yaml`.

Comment (yike21) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/resourceinterpretercustomization.yaml:1]: Sorry, I will fix the file name soon.

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    statusAggregation:
+      luaScript: >
+        function AggregateStatus(desiredObj, statusItems)
+          if statusItems == nil then
+            return desiredObj
+          end
+          if desiredObj.status == nil then
+            desiredObj.status = {}
+          end
+          if desiredObj.generation == nil then
```
[the remark this answers is not in the document]
Comment (yike21) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:33]: > ```diff
> -          if desiredObj.generation == nil then
> -            desiredObj.generation = 0
> +          if desiredObj.metadata.generation == nil then
> +            desiredObj.metadata.generation = 0
>            end
> -          generation = desiredObj.generation
> +          generation = desiredObj.metadata.generation
>            replicas = 0
> ```

Thanks a lot! 
This also explains why the previous interpretHealth section would not pass when it came to the generation field, and I will fix it and the `interpretHealth` rule. [related-question2](https://github.com/karmada-io/karmada/issues/3331#issuecomment-1492990755)

[the remark this answers is not in the document]
Comment (yike21) [reply pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml]: karmadactl interpret validation is OK, and the `healthInterpretation` rule is similar to native deployment. [link](https://github.com/karmada-io/karmada/blob/master/pkg/resourceinterpreter/default/native/healthy.go#L31)

File: pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+          status.readyReplicas = observedObj.status.readyReplicas
+          status.availableReplicas = observedObj.status.availableReplicas
+          status.updatedReadyReplicas = observedObj.status.updatedReadyReplicas
+          status.expectedUpdatedReplicas = observedObj.status.expectedUpdatedReplicas
+          status.updateRevision = observedObj.status.updateRevision
+          status.currentRevision = observedObj.status.currentRevision
+          status.observedGeneration = observedObj.status.observedGeneration
+          return status
+        end
+    healthInterpretation:
```
Comment (yike21) [inline pkg/resourceinterpreter/default/thirdparty/resourcecustomizations/apps.kruise.io/v1alpha1/CloneSet/customizations.yaml:103]: The `healthInterpretation` rule for `cloneset` is a reference to the code in `pkg/resourceinterpreter/default/native/healthy.go` corresponding to native deployment. The idea is changed from "determine what is healthy" to "determine what is unhealthy/illegal". It allows a member cluster to be considered healthy when the pod is not fully ready, and to be aggregated.

**Pull request merged**